File: 1411-number-of-ways-to-paint-n-3-grid/1411-number-of-ways-to-paint-n-3-grid.cpp
class Solution {
public:
    int numOfWays(int n) {
        const long long MOD = 1e9 + 7;
        long long aba = 6;
        long long abc = 6;

        for (auto i = 1; i<n; i++){
       long long new_aba = (aba * 3 + abc * 2) % MOD;
            long long new_abc = (aba * 2 + abc * 2) % MOD;
            
            aba = new_aba;
            abc = new_abc;
        }
        
        return (aba + abc) % MOD;
    }
};

File: 1411-number-of-ways-to-paint-n-3-grid/1411-number-of-ways-to-paint-n-3-grid.rs
impl Solution {
    pub fn num_of_ways(n: i32) -> i32 {
        const MOD: i64 = 1_000_000_007;
        let mut aba: i64  = 6;
        let mut abc: i64  = 6;
         for _ in 1..n {
            let new_aba = (aba * 3 + abc * 2) % MOD;
            let new_abc = (aba * 2 + abc * 2) % MOD;
            
            aba = new_aba;
            abc = new_abc;
        }
        
        ((aba + abc) % MOD) as i32
    }
}